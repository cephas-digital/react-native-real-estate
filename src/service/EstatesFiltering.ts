import { Estate, ListingFilters } from "./../interfaces";

export const filterEstatesBy = (
  estates: Array<Estate>,
  filters: ListingFilters
): Array<Estate> => {
  estates = estates
    .map(
      (estate: Estate): Estate => {
        estate.hidden = false;
        return estate;
      }
    )
    .map(
      (estate: Estate): Estate => {
        estate.hidden = false;

        if (
          estate.city.country !== filters.country &&
          filters.country !== null &&
          typeof filters.country !== "undefined"
        ) {
          estate.hidden = true;
        }

        return estate;
      }
    )
    .filter((estate: Estate): boolean => !estate.hidden)
    .map(
      (estate: Estate): Estate => {
        if (
          estate.isOnSale !== filters.onSale &&
          filters.onSale !== 2 &&
          filters.onSale !== null &&
          typeof filters.onSale !== "undefined"
        ) {
          estate.hidden = true;
        }

        return estate;
      }
    )
    .filter((estate: Estate): boolean => !estate.hidden)
    .map(
      (estate: Estate): Estate => {
        if (
          filters.maxPrice !== null &&
          typeof filters.maxPrice !== "undefined" &&
          estate.price > filters.maxPrice
        ) {
          estate.hidden = true;
        }

        return estate;
      }
    );

  return estates;
};
