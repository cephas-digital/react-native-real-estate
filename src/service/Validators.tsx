import { View, Text } from "react-native";
import React, { Fragment } from "react";
import { NewEstate } from "../interfaces";

export const validateNewEstate = (newEstate: NewEstate) => {
  let hasErrors = false;
  const errors: Array<Element> = [];
  const errorMessages = [];
  const { city_id, developer_id, price, name } = newEstate;

  if (city_id === null || !Number.isInteger(city_id)) {
    hasErrors = true;
    errorMessages.push("Selected country is invalid!");
  }

  if (
    developer_id === null ||
    typeof developer_id === "undefined" ||
    !Number.isInteger(developer_id)
  ) {
    hasErrors = true;
    errorMessages.push("Selected developer is invalid!");
  }

  if (
    price === null ||
    typeof price === "undefined" ||
    !Number.isInteger(price)
  ) {
    hasErrors = true;
    errorMessages.push("Provided price is invalid!");
  }

  if (name === null || name.length < 1) {
    hasErrors = true;
    errorMessages.push("Provided name is invalid!");
  }

  errorMessages.forEach((errorMessage, index) => {
    errors.push(
      <View
        key={index}
        style={{
          backgroundColor: "#ff5722",
          marginVertical: 4,
          borderRadius: 4,
          borderWidth: 1,
          borderColor: "#ff5722"
        }}
      >
        <Text style={{ color: "#fff", padding: 4 }}>{errorMessage}</Text>
      </View>
    );
  });

  return { hasErrors, errors: <Fragment>{errors}</Fragment> };
};
