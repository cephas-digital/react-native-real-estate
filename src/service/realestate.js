const fetch = require('node-fetch');

const url = (path) => (process.env.URL || ('http://laravel-task.herokuapp.com/api/')) + path;

function buildParamsQuery(params) {
  const esc = encodeURIComponent;
  return Object.keys(params)
    .map(k => esc(k) + '=' + esc(params[k]))
    .join('&');
}

module.exports = class RealEstateApi {
  static getList() {
    return fetch(url('lists'))
      .then(res => res.json());
  }

  static getCities() {
    return fetch(url('cities'))
      .then(res => res.json());
  }

  static getDevelopers() {
    return fetch(url('developers'))
      .then(res => res.json());
  }

  static filterByType(data) {
    console.log(url('estate/filter?' + buildParamsQuery(data)))
    return fetch(url('estate/filter?' + buildParamsQuery(data)))
      .then(res => res.json())
  }

  static getById(id) {
    return fetch(url(`estate/${id}`))
      .then(res => res.json())
  }

  static addListItem(data) {
    return fetch(url(`estate/add`), {
      headers: {
        'Accept': 'application/json',
        'Content-Type': 'application/json'
      },
      method: 'post',
      body: JSON.stringify({
        cityId: data.cityId,
        developerId: data.developerId,
        name: data.name,
        onSale: data.onSale,
        price: data.lessThen
      })
    })
    .then(res => res.json())
  }
}