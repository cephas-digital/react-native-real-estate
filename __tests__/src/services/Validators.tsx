import React from "react";
import { validateNewEstate } from "../../../src/service/Validators";
import renderer from "react-test-renderer";

test("renders correctly", () => {
  const validationResult = validateNewEstate({
    name: null,
    price: null,
    city_id: null,
    developer_id: null
  });

  expect(validationResult.hasErrors).toBeTruthy();
  const tree = renderer.create(validationResult.errors).toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders single error", () => {
  const validationResult = validateNewEstate({
    name: "test estate",
    price: 15000,
    city_id: 12,
    developer_id: null
  });

  expect(validationResult.hasErrors).toBeTruthy();
  const tree = renderer.create(validationResult.errors).toJSON();
  expect(tree).toMatchSnapshot();
});

test("renders price error", () => {
  const validationResult = validateNewEstate({
    name: "test estate",
    price: "sdfsd",
    city_id: 12,
    developer_id: 32
  });

  expect(validationResult.hasErrors).toBeTruthy();
  const tree = renderer.create(validationResult.errors).toJSON();
  expect(tree).toMatchSnapshot();
});

test("no errors", () => {
  const validationResult = validateNewEstate({
    name: "test estate",
    price: 1400,
    city_id: 12,
    developer_id: 32
  });

  expect(validationResult.hasErrors).toBeFalsy();
  const tree = renderer.create(validationResult.errors).toJSON();
  expect(tree).toMatchSnapshot();
});
