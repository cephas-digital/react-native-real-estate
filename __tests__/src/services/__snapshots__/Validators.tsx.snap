// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no errors 1`] = `null`;

exports[`renders correctly 1`] = `
Array [
  <View
    style={
      Object {
        "backgroundColor": "#ff5722",
        "borderColor": "#ff5722",
        "borderRadius": 4,
        "borderWidth": 1,
        "marginVertical": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "padding": 4,
        }
      }
    >
      Selected country is invalid!
    </Text>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "#ff5722",
        "borderColor": "#ff5722",
        "borderRadius": 4,
        "borderWidth": 1,
        "marginVertical": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "padding": 4,
        }
      }
    >
      Selected developer is invalid!
    </Text>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "#ff5722",
        "borderColor": "#ff5722",
        "borderRadius": 4,
        "borderWidth": 1,
        "marginVertical": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "padding": 4,
        }
      }
    >
      Provided price is invalid!
    </Text>
  </View>,
  <View
    style={
      Object {
        "backgroundColor": "#ff5722",
        "borderColor": "#ff5722",
        "borderRadius": 4,
        "borderWidth": 1,
        "marginVertical": 4,
      }
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "padding": 4,
        }
      }
    >
      Provided name is invalid!
    </Text>
  </View>,
]
`;

exports[`renders price error 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ff5722",
      "borderColor": "#ff5722",
      "borderRadius": 4,
      "borderWidth": 1,
      "marginVertical": 4,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
        "padding": 4,
      }
    }
  >
    Provided price is invalid!
  </Text>
</View>
`;

exports[`renders single error 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#ff5722",
      "borderColor": "#ff5722",
      "borderRadius": 4,
      "borderWidth": 1,
      "marginVertical": 4,
    }
  }
>
  <Text
    style={
      Object {
        "color": "#fff",
        "padding": 4,
      }
    }
  >
    Selected developer is invalid!
  </Text>
</View>
`;
